feat: Add knowledge bank entry generation from human replies

ConstantTime opened the pull request:
**Closes:** https://github.com/antiwork/helper/issues/530

## Problem

When human agents send valuable replies containing step-by-step instructions, troubleshooting solutions, or frequently asked information, this knowledge was not being captured for future reuse. Agents would often provide excellent answers that could benefit other customers, but these insights remained buried in individual conversation threads.

## Solution

Added an intelligent knowledge bank suggestion system that analyzes human replies and offers to create reusable knowledge base entries. The system:

- **Smart Analysis**: AI determines when human replies contain valuable, reusable information vs. customer-specific responses
- **One-Click Access**: Added a "💡 KB" button to the message sent toast for easy access
- **Guided Creation**: Opens an interactive dialog with AI-generated suggestions that can be edited and refined
- **Seamless Integration**: Works with existing knowledge bank infrastructure and embedding system

The AI intelligently filters out customer-specific details and extracts generalizable information suitable for a knowledge base, ensuring only valuable content becomes reusable entries.

## Recordings

### Basic Flow Demo

**How a helpful reply is saved.**

https://github.com/user-attachments/assets/e9554893-52e6-4e7f-a006-8bf1b61ac351

### AI Intelligence Demo  

**How a non-helpful reply is ignored**

https://github.com/user-attachments/assets/65b6cd6f-6433-4334-b0df-6d9beac721fe

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- **New Features**
  - Added the ability to generate knowledge bank entries directly from sent messages via a new dialog interface.
  - Introduced a "💡 Save" action in message notifications to quickly open the knowledge bank generation dialog.
  - Implemented AI-powered suggestions for knowledge bank entries based on human replies, including editable content and AI reasoning.

- **User Experience**
  - Enhanced feedback with toast notifications for successful or failed knowledge bank entry generation.
  - Improved workflow for saving and editing suggested knowledge bank content.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

ConstantTime commented on app/(dashboard)/mailboxes/[mailbox_slug]/[category]/conversation/messageActions.tsx:
```diff
@@ -221,6 +227,15 @@ export const MessageActions = () => {
+                setShowKnowledgeBankDialog(true);
+              }}
+            >
+              💡 KB
```
The toast design was getting messed up hence made this shorter. Checking what can be done here.

ConstantTime replied:
How about this?

<img width="290" alt="Screenshot 2025-06-23 at 7 13 59 PM" src="https://github.com/user-attachments/assets/97e79ecb-0e52-4bde-b168-e76fdc7444d0" />

ConstantTime replied:
But do we only want the last reply from agent to be saved there? I think the current UX provides flexibility into what the agent wants to save / thinks would be helpful enough to save?

slavingia replied:
It would be the same in terms of running the prompt and opening the modal for user input, just makes it more front and center. This is fine for now though since we already have the toast imo

ConstantTime commented:
Latest demo

https://github.com/user-attachments/assets/b8681182-310c-4a8c-ab45-78bf6500cc29

binary-koan commented on trpc/router/mailbox/faqs.ts:
```diff
@@ -135,4 +143,79 @@ export const faqsRouter = {
+
+      const similarFAQs = await findEnabledKnowledgeBankEntries(ctx.mailbox);
+
+      const systemPrompt = `
```
Ah I just realised `update_entry` isn't here. Is there a reason for that? Ideally we don't want it creating a bunch of similar entries for subtly different cases (or when existing entries are out of date)

ConstantTime replied:
hmm, fair ask. Implementing this. There is a check for exactly same entries but I think we can avoid similar entries too via LLMs:

https://github.com/user-attachments/assets/ccb5ea69-ff0d-43dd-adca-57ab6cdc6b03

ConstantTime replied:
Yes, [updated now](https://github.com/antiwork/helper/pull/567#issuecomment-2996695045) after introducing update also.

ConstantTime commented:
> ### Bug: AI Suggestion Flow Ignores Pending Entries
> The `generateKnowledgeBankSuggestion` function incorrectly provides only enabled knowledge bank entries to the AI, omitting existing pending suggestions. This causes the AI to be unaware of already suggested entries, leading to the creation of duplicate suggestions and missed opportunities to update existing ones, resulting in inconsistent behavior compared to previous flows.
> 
> `lib/ai/knowledgeBankSuggestions.ts#L24-L82`
> [Fix in Cursor](https://cursor.com/open?data=eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6ImJ1Z2JvdC12MSJ9.eyJ2ZXJzaW9uIjoxLCJ0eXBlIjoiQlVHQk9UX0ZJWF9JTl9DVVJTT1IiLCJkYXRhIjp7InJlZGlzS2V5IjoiYnVnYm90OjllZmQzNGFhLTM3NmEtNDgyMi04Y2Y1LTZhYTQ4NmE0Njg0MiIsImVuY3J5cHRpb25LZXkiOiJHRkRmVUkwRmN1WVFTOUtYRm9VUUhUZVZCVFc5Ml9kU0l1dEFsLXN2RWJzIiwiYnJhbmNoIjoiZmVhdC9rbm93bGVkZ2UtYmFuay1nZW5lcmF0aW9uLWh1bWFuLXJlcGxpZXMifSwiaWF0IjoxNzUwNjg4NTk2LCJleHAiOjE3NTEyOTMzOTZ9.efhpYQR_D9pUdG_EUYQERI1ERQAiHpocWHcWrft03CZkPBpf3Rwm9tjrDpvWTP--wT6cOze3sMcaQ0uDA0SQt8qR3ZApDmTgAHBiMXWf0tuN7kel3eNv96pcC46k0um1tMfINCPhkHuByC1t4LhSGzHCyGkTO7NAnmi55Ty4lEybsPPCq-QNffRKRE5U--5_KGmOrRfxA9u-2DYN1La3pnuKfUIR6i-tRjAI_pWQGdP-yR5Kg0e2g33ykk3HtZ-DG9xHiLVo3ADAxJZglLVr5ixc8MXOFwwuIxVLRb3mvsoZpLq_VZRLQcUTEZ5AZ6w4Bu-41ugjjCo54IEf1-sjkA)
> 
> _Was this report helpful? Give feedback by reacting with 👍 or 👎_

hmm good catch but I think this would be quite a bit of over engineering? will have to save pending ones also somewhere * _make sure they are deleted also timely?_

slavingia commented:
Yeah not with fixing

binary-koan commented on trpc/router/mailbox/faqs.ts:
```diff
@@ -135,4 +136,42 @@ export const faqsRouter = {
+        },
+      });
+
+      if (!message) {
+        throw new TRPCError({ code: "NOT_FOUND", message: "Message not found" });
+      }
+
+      if (message.conversation.mailboxId !== ctx.mailbox.id) {
```
```suggestion
      if (message?.conversation.mailboxId !== ctx.mailbox.id) {
```

might be enough?

binary-koan reviewed (commented):
Added a few more minor comments, but overall this is looking great!

ConstantTime commented:
Addressed comments @binary-koan. Screenshots after the suggested changes.

**New UI for update KB flow**

<img width="626" alt="Screenshot 2025-06-24 at 4 10 34 PM" src="https://github.com/user-attachments/assets/443119a1-56c3-4022-8c24-546411ae1d4e" />

**No-op toast**

<img width="356" alt="Screenshot 2025-06-24 at 4 11 14 PM" src="https://github.com/user-attachments/assets/fef5a7ba-420b-45a7-82eb-ac28b13fcf8d" />

binary-koan commented:
I think we should put the toast actions below the title instead of beside and then rename "save" to "generate knowledge", but will do that separately
